x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -306,9 +355,49 @@ public void restoreShard(IndexShard indexShard, SnapshotId snapshotId, Version v
             updateMappings(remoteClient, leaderIndex, restoreSession.mappingVersion, client, indexShard.routingEntry().index());
         } catch (Exception e) {
             throw new IndexShardRestoreFailedException(indexShard.shardId(), "failed to restore snapshot [" + snapshotId + "]", e);
+        } finally {
+            logger.trace("{} stopping background retention lease renewing at the end of recovery", shardId);
+            renewable.cancel();
         }
     }
 
+    private Optional<RetentionLeaseAlreadyExistsException> addRetentionLease(
+            final ShardId leaderShardId,
+            final String retentionLeaseId,
+            final Client remoteClient) {
+        try {
+            final RetentionLeaseActions.AddRequest request =
+                    new RetentionLeaseActions.AddRequest(leaderShardId, retentionLeaseId, RETAIN_ALL, "ccr");
+            remoteClient.execute(RetentionLeaseActions.Add.INSTANCE, request).actionGet(ccrSettings.getRecoveryActionTimeout());
+            return Optional.empty();
+        } catch (final RetentionLeaseAlreadyExistsException e) {
+            return Optional.of(e);
+        }
+    }
+
+    private Optional<RetentionLeaseNotFoundException> renewRetentionLease(
+            final ShardId leaderShardId,
+            final String retentionLeaseId,
+            final Client remoteClient) {
+        try {
+            final RetentionLeaseActions.RenewRequest request =
+                    new RetentionLeaseActions.RenewRequest(leaderShardId, retentionLeaseId, RETAIN_ALL, "ccr");
+            remoteClient.execute(RetentionLeaseActions.Renew.INSTANCE, request).actionGet(ccrSettings.getRecoveryActionTimeout());
+            return Optional.empty();
+        } catch (final RetentionLeaseNotFoundException e) {
+            return Optional.of(e);
+        }
+    }
+
+    // this setting is intentionally not registered, it is only used in tests
+    public static final Setting<TimeValue> RETENTION_LEASE_RENEW_INTERVAL_SETTING =
+            Setting.timeSetting(
+                    "index.ccr.retention_lease.renew_interval",
```

Comment:
if this is only used for recovery from remote, perhaps add `recovery` somewhere to the name

Reply:
This setting will end up being shifted in the follow-on PR to also be used for the renewal interval while following. Therefore I think recovery should stay out of the name, and I will merely lift this setting up higher when ready.